devito/finite_differences/derivative.py

```diff
@@ -267,7 +268,14 @@ def _eval_at(self, func):
         setup where one could have Eq(u(x + h_x/2), v(x).dx)) in which case v(x).dx
         has to be computed at x=x + h_x/2.
         """
-        x0 = dict(zip(func.dimensions, func.indices_ref))
+        # Split the Add case to handle different staggereing
+        x0 = dict(func.indices_ref._getters)
+        if self.expr.is_Add:
```

Comment:
I have a feeling that anything resorting to `is_Add` or `is_Mul` during evaluation is breakable more or less easily

Reply:
`x*x` isn't a `Mul`, but it's a multiplication
`cos(x)` is neither addition nor multiplication

I mean, I admit I haven't thought much about this, but it doesn't sound very robust after a first glance

Reply:
-`x*x` has no issue has they both have smae indices and fallback to default works perfectly.
- `cos(x)` has zero issue as `cos` doesn't have indices so `cos(x)` and `x` have smae indices and fallback works perfectly

Reply:
>is breakable more or less easily

why? All inputs are expressions (and is checked at creation), if not would break with or without this.

Reply:
I find it difficult to believe that it boils down to just acting differently based on whether `expr` is an Add or Mul. So, you're associating semantics to the type of operation, do I get it right? 

and if it's neither a Mul nor an Add, then we're 100% sure just returning `self._new_from_self(x0=x0)` is _always_ correct?

Reply:
Yes it is as all non `Add/Mul` are just functions (cos/sin/..) that are unary and don't requires special handling (except the open issue on having our own versions of these functions).

Reply:
is it worth adding

```
else:
     assert len(self.expr.args) <= 1
```

?

Reply:
Why?

Reply:
to assert this:
> Yes it is as all non Add/Mul are just functions (cos/sin/..) that are unary

perhaps nitpicking . Just a bit concerned about us forgetting about something or SymPy changing over time. You can probably ignore this

Reply:
My brain is a bit shot atm, but this will ok if `expr` is and `Add` object with embeded `Mul`'s etc right? (i.e. `expr=f*g+f*h`).

Reply:
Well not really. Currently only pure Add `a+b` and Mul `a*b` are going through this. For more complex expression, the assumption is that there is probably at least two different objects with the same priority and can't be safely shifted so should stay as is. I have some ideas for these cases but will need major refactor of a few pieces.

Reply:
> Well not really. 

ahhhhh...that explains it. That was a big source of confusion for me, maybe I didn't explain it appropriately. 

>  I have some ideas for these cases but will need major refactor of a few pieces.

If addressing the general case would need a major refactor, then I suggest we at least:

* add a giant TODO here in the code
* add an issue to GH
* an x-failing test

then we can proceed I think

Reply:
No, this is not a TODO or anything, this is not possible to support these kind of expression without hardocding hacks, allI am saying is that there is ways for specific cases but not here, so these would just  make thing worse pretending there is a way to do it..

Reply:
I still don't understand what's the semantic information encoded in `Add` , IOW, what `Add` represents to require a different treatment than `Mul`. But yeah, maybe it's just me being slow, maybe you can elaborate offline? I think I'm quite curious to understand at this point

Reply:
cool now